tests/unittests/test_length_mask.py

```diff
@@ -0,0 +1,598 @@
+) -> Dict[int, List[int]]:
+    """
+    Compute fractions that where numerical accuracy can be an issue for
+    fractional lengths.
+
+    Arguments
+    ---------
+    device: torch.device
+        Torch device for which the fractions are adversarial.
+    dtype: torch.dtype
+        Torch dtype for which the fractions are adversarial.
+    max_value: int
+        Maximum value (inclusive) of denominator and numerator to be considered.
+
+    Returns
+    -------
+    A dictionary from d to a list of n, where (n/d) * d != n when computed by
+    Torch with the device and dtype passed in, and n < d.
+    """
+    max_value = 32
+    indices = torch.arange(1, max_value + 1, dtype=dtype, device=device)
+    numerator = torch.broadcast_to(indices, (max_value, max_value))
+    denominator = torch.transpose(numerator, 0, 1)
+    less_than_one = numerator < denominator
+    inexact = ((numerator / denominator) * denominator) != numerator
+
+    result = defaultdict(list)
+    for n, d in zip(
+        numerator[less_than_one * inexact],
+        denominator[less_than_one * inexact],
+    ):
+        result[int(d)].append(int(n))
+    return result
+
+
+def generate_integer_lengths(
+    generator: np.random.Generator,
+    lengths_ndim: int,
+    adversarial_device: Optional[torch.device] = None,
+    fractional_dtype: Optional[torch.dtype] = None,
+) -> Iterable[Tuple[int, np.ndarray, np.ndarray]]:
+    """
+    Generate example lengths that are integer, as well as their fractional
+    counterparts.
+    Most weights are randomly chosen, but some are chosen so that the fractional
+    versions, when multiplied back, are not exactly integer.
+
+    Arguments
+    ---------
+    generator: np.random.Generator
+        The Numpy generator that ensures deterministic examples.
+    lengths_ndim: int
+        The number of dimensions that the length array should have.
+    adversarial_device: Optional[torch.device]
+        The Torch device used for finding lengths that are rounded
+        incorrectly.
+    fractional_dtype: Optional[torch.dtype]
+        The Torch dtype used for checking rounding.
+
+    Yields
+    ------
+    Examples (max_length, integer_lengths, fractional_lengths) where the
+    fractional_lengths match integer_lengths.
+    """
+
+    def random_lengths(max_length=None):
+        """Draw random lengths."""
+        if max_length is None:
+            max_length = int(generator.integers(low=1, high=10))
+        lengths_shape = generator.integers(low=1, high=8, size=lengths_ndim)
+        lengths = generator.integers(
+            low=0, high=max_length + 1, size=lengths_shape
+        )
+        return max_length, lengths
+
+    def generate_lengths():
+        """Generate first adversarial and then random weights."""
+        if adversarial_device is not None:
+            assert fractional_dtype is not None
+            deterministic_lengths = adversarial_fractions(
+                adversarial_device, fractional_dtype, 32
```

Comment:
since you are hardcoding `max_value` to `32` here as well, perhaps leaving it to `32` directly as a default parameter of your function would be better?